engine/table/src/main/java/io/deephaven/engine/table/impl/InstrumentedTableListenerBase.java

```diff
@@ -152,9 +152,12 @@ public boolean satisfied(final long step) {
 
     @Override
     public void onFailure(Throwable originalException, Entry sourceEntry) {
+        deregisterOnFailure();
         onFailureInternal(originalException, sourceEntry == null ? entry : sourceEntry);
     }
 
+    protected abstract void deregisterOnFailure();
```

Comment:
I think this current implementation is a big improvement over the status quo. I'm trying to think if we could go one better and drop the reference count to zero, thus forcing a destroy. Maybe this whole change set belongs in a separate PR.